Production/scripts/validation_tool.py

```diff
@@ -0,0 +1,159 @@
+  histos[0].SetMarkerStyle(20)
+
+  for ii, hh in enumerate(histos): 
+    hh.SetLineColor(ii+1)
+    hh.SetMarkerColor(ii+1)
+    leg.AddEntry(hh, 'chunk %d - pval = %.3f' %(ii, pvalues[ii]), 'lep')
+    hh.Draw("PE" + " SAME" * (ii != 0))
+    hh.Write()
+  if args.legend:
+    leg.Draw("SAME")
+  if args.pdf is not None:
+    can.SaveAs('{}/{}.pdf'.format(args.pdf, fdir.replace('/', '_')), 'pdf')
+  can.Write()
+  OUTPUT_ROOT.cd()
+
+def run_validation(dataframe, branches, pwd):
+  OUTPUT_ROOT.cd()
+  OUTPUT_ROOT.mkdir(pwd)
+  
+  if not pwd in JSON_DICT.keys():
+    JSON_DICT[pwd] = OrderedDict()
+
+  for branch in branches:
+    OUTPUT_ROOT.cd() 
+    OUTPUT_ROOT.mkdir('/'.join([pwd, branch]))
+
+    histos = get_histos(dataframe, branch = branch, norm = True)
+
+    pvalues = [histos[0].KolmogorovTest(hh) if histos[0].Integral()*hh.Integral() else 99 for hh in histos]
+    if not histos[0].Integral():
+      print '[WARNING] control histogram is empty for step {} inside {}'.format(branch, pwd)
+    
+    if not all([pv >= PVAL_THRESHOLD for pv in pvalues]):
+      print '[WARNING] KS test failed for step {} inside {}. p-values are:'.format(branch, pwd)
+      print '\t', pvalues
+    
+    JSON_DICT[pwd][branch] = pvalues
+    
+    save_histos(histos, fdir = '/'.join([pwd, branch]), pvalues = pvalues)
+
+if __name__ == '__main__':
+  print '[INFO] reading files', args.input
+  input_files = ROOT.std.vector('std::string')()
+  for file in glob.glob(args.input):
+    input_files.push_back(str(file))
+
+  main_dir = 'KS_test'
+
+  ## first, run on plain columns
+  dataframe = ROOT.RDataFrame('taus', input_files)
+  run_validation(dataframe = dataframe, pwd = main_dir, branches = ['lepton_gen_match', 'sampleType', 'dataset_group_id', 'dataset_id'])
+
+  ## then, group by tau type
+  tau_type_dataframes = groupby(dataframe = dataframe, by = 'lepton_gen_match')
+  for ii, df in tau_type_dataframes.iteritems():
+    run_validation(dataframe = df, pwd = '/'.join([main_dir, 'lepton_gen_match', str(ii)]), branches = ['tau_pt', 'tau_eta'])
+
+  ## then, group by sample type
+  sample_type_dataframes = groupby(dataframe = dataframe, by = 'sampleType')
+  for ii, df in sample_type_dataframes.iteritems():
+    run_validation(dataframe = df, pwd = '/'.join([main_dir, 'sampleType', str(ii)]), branches = ['tau_pt', 'tau_eta'])
+
+  ## then, group by dataset group id
+  group_id_dataframes = groupby(dataframe = dataframe, by = 'dataset_group_id')
+  for ii, df in group_id_dataframes.iteritems():
+    run_validation(dataframe = df, pwd = '/'.join([main_dir, 'dataset_group_id', str(ii)]), branches = ['tau_pt', 'tau_eta', 'dataset_id'])
+
+  ## then, group by dataset id
+  group_id_dataframes = groupby(dataframe = dataframe, by = 'dataset_id')
+  for ii, df in group_id_dataframes.iteritems():
+    run_validation(dataframe = df, pwd = '/'.join([main_dir, 'dataset_id', str(ii)]), branches = ['tau_pt', 'tau_eta'])
```

Comment:
it will be faster if `run_validation` is split into two parts:
- first, get histogram pointers for all data frames and all branches
- then do another loop over data frames and branches, and transform pointers into histograms by calling GetValue.

In this way, it would require only one passage over the input files. For a large dataset, multiple passages would considerably slow down the processing time.

Reply:
I think I can work something out using filters instead of ranges,

Reply:
26 times seems a bit too much. I agree that we can implement things without ranges then (and, as a bonus, we'll be able to run with implicit multithreading). Anyhow, we need to process the entire dataset in order to do validation, so probably ranges (in their current implementation) is not something we want.
I don't think you need filters for that:
- define variable chunk_id: `df = df.Define("chunk_id", "rdfentry_ % {}".format(n_chunks))`
- create 2D histograms instead 1D, where the second dimension is filled with chunk_id (for pt/eta it will be 3D histogram then)
- After 2D histograms are created and filled, convert each slice into 1D and the rest is equivalent to how it is implemented currently in v2.

Reply:
Sorry, small correction for chunk_id definition:
```python
df = df.Define("chunk_id", "rdfentry_ * {} / {}".format(n_chunks, total_n_entries))
```

Reply:
I've updated the [validation tool](https://github.com/lguzzi/TauMLTools/blob/dev/Production/scripts/validation_tool-v5.py) in my dev branch. Before merging to master, could you please explain why shouldn't I use 
```
df = df.Define("chunk_id", "rdfentry_ % {}".format(n_chunks))
```

Reply:
With `%` your chunks will be distributed through the whole dataset. E.g. with two chunks you'll be testing compatibility between even and odd entries. It is not what we want. We want to check compatibility between different parts of the dataset. E.g. with two chunks we want to test that the first 50% of taus having the same properties as the last 50% of taus.
Typical batch size is about 100-200 taus, so even if by some mistake taus with even entry id would have different properties wrt to the taus with odd entry id - it will be transparent for the training. On the other hand, if the first half of the dataset has different properties, compared to the second part of the dataset - it is a problem: the likelihood minimums will be different in different parts => minimization will not be able to converge.

Reply:
Thank you, I've updated the script with your suggestions